### PR TITLE
scipy: register conflict with too-recent openblas

### PR DESCRIPTION
**Pull request by liam-o-marsh**
(on another note, a dependency bound on ~~numpy~~ EDIT: PYTHON was made more strict after looking at the dependency matrix on scipy's own online documentation)

**assumtions made by this patch:**:
- version ranges (like @2.1:2.2) are inclusive (I'm mostly sure of this, but I'ld rather not submit a broken patch if this is wrong)

checks:
- [X] spack unit-test -> no worse than before (4670 passed, 125 skipped, 11 xfailed, 18 xpassed, 40 warnings)
- [X] spack style -> OK
- [X] spec correctly constructed for most recent scipy
- [ ] scipy tests (I am not sure how to run them correctly on a pre-installed scipy. `spack test run py-scipy` seems to only return errors linked to not having cupy and torch installed)

### CODE REVIEW

**Comment by adamjstewart** [comment]
Yes, version ranges are inclusive.

You can run the scipy tests using `spack install --test=root py-scipy`. Not sure why `spack test run py-scipy` isn't working for you.

**Comment by adamjstewart** [comment]
Partial duplicate of #42213. I would suggest making this PR only about openblas conflicts and not about adding a new version.

**Comment by liam-o-marsh** [comment]
alright, it is no longer a partial duplicate.
I'll run the tests now, and see what happens.

**Comment by liam-o-marsh** [comment]
`spack install --test=root py-scipy` didn't run any tests, but I found out about `python3 -c 'import scipy;scipy.test()'`
<details>
<summary>[this is a dropdown] `= 40 failed, 42452 passed, 2806 skipped, 12693 deselected, 136 xfailed, 10 xpassed in 533.98s (0:08:53) =`</summary>
```
FAILED scipy/io/matlab/tests/test_streams.py::TestZlibInputStream::test_all_data_read_overlap
FAILED scipy/io/matlab/tests/test_streams.py::TestZlibInputStream::test_all_data_read_bad_checksum
FAILED scipy/linalg/tests/test_special_matrices.py::test_special_matrices_deprecation[tri]
FAILED scipy/optimize/tests/test_optimize.py::TestBrute::test_runtime_warning
FAILED scipy/signal/tests/test_filter_design.py::TestButtord::test_runtime_warnings
FAILED scipy/signal/tests/test_signaltools.py::TestFFTConvolve::test_fft_nan
FAILED scipy/signal/tests/test_signaltools.py::TestMedFilt::test_none - Depre...
FAILED scipy/signal/tests/test_signaltools.py::TestMedFilt::test_refcounting
FAILED scipy/signal/tests/test_windows.py::test_deprecated_import[chebwin-args15]
FAILED scipy/sparse/linalg/_eigen/lobpcg/tests/test_lobpcg.py::test_ElasticRod
FAILED scipy/sparse/linalg/_eigen/lobpcg/tests/test_lobpcg.py::test_failure_to_run_iterations_nonsymmetric
FAILED scipy/sparse/linalg/_eigen/lobpcg/tests/test_lobpcg.py::test_verbosity
FAILED scipy/sparse/linalg/_eigen/lobpcg/tests/test_lobpcg.py::test_maxit - U...
FAILED scipy/sparse/linalg/_isolve/tests/test_iterative.py::TestGMRES::test_restrt_dep
FAILED scipy/sparse/tests/test_base.py::TestCOO::test_multiple_ellipsis_slicing
FAILED scipy/sparse/tests/test_base.py::TestDIA::test_multiple_ellipsis_slicing
FAILED scipy/sparse/tests/test_base.py::TestBSR::test_multiple_ellipsis_slicing
FAILED scipy/sparse/tests/test_base.py::TestBSRNonCanonical::test_multiple_ellipsis_slicing
FAILED scipy/sparse/tests/test_base.py::TestCOONonCanonical::test_multiple_ellipsis_slicing
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_limit_10[TestBSR-test_multiple_ellipsis_slicing]
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_limit_10[TestCOO-test_multiple_ellipsis_slicing]
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_limit_10[TestDIA-test_multiple_ellipsis_slicing]
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_random[TestBSR-test_multiple_ellipsis_slicing]
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_random[TestCOO-test_multiple_ellipsis_slicing]
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_random[TestDIA-test_multiple_ellipsis_slicing]
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_all_32[TestBSR-test_multiple_ellipsis_slicing]
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_all_32[TestCOO-test_multiple_ellipsis_slicing]
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_all_32[TestDIA-test_multiple_ellipsis_slicing]
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_all_64[TestBSR-test_multiple_ellipsis_slicing]
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_all_64[TestCOO-test_multiple_ellipsis_slicing]
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_all_64[TestDIA-test_multiple_ellipsis_slicing]
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_no_64[TestBSR-test_multiple_ellipsis_slicing]
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_no_64[TestCOO-test_multiple_ellipsis_slicing]
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_no_64[TestDIA-test_multiple_ellipsis_slicing]
FAILED scipy/special/tests/test_cython_special.py::test_cython_api[spherical_kn]
FAILED scipy/stats/tests/test_fit.py::TestFit::test_guess_iv - RuntimeWarning...
FAILED scipy/stats/tests/test_mstats_basic.py::test_theilslopes_warnings - Ru...
FAILED scipy/stats/tests/test_stats.py::TestMoments::test_moment - RuntimeWar...
FAILED scipy/stats/tests/test_stats.py::TestMoments::test_empty_1d - RuntimeW...
FAILED scipy/stats/tests/test_stats.py::TestBrunnerMunzel::test_brunnermunzel_return_nan
```
</details>

...not sure which ones of those are 'expected' and which should require me to go find other conflicts. I thin I'll try with a slightly older numpy, see if some of those disappear.
[edit: it only removes one failure, which was basically "whoops the wrong deprecation warning was thrown"]

File: var/spack/repos/builtin/packages/py-scipy/package.py [hunk 1]
```diff
@@ -50,7 +50,7 @@ class PyScipy(PythonPackage):
     version("1.3.2", sha256="a03939b431994289f39373c57bbe452974a7da724ae7f9620a1beee575434da4")
 
     # Based on wheel availability on PyPI
-    depends_on("python@3.9:3.12", when="@1.11:", type=("build", "link", "run"))
+    depends_on("python@3.9:3.11", when="@1.11:", type=("build", "link", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-scipy/package.py]: Can you undo these changes too? I have them in the other PR

File: var/spack/repos/builtin/packages/py-scipy/package.py [hunk 2]
```diff
@@ -90,7 +90,8 @@ class PyScipy(PythonPackage):
     depends_on("py-setuptools@:59", when="@1.8", type="build")
     depends_on("py-setuptools@:57", when="@1.7", type="build")
     depends_on("py-setuptools@:51.0.0", when="@1.6", type="build")
-    depends_on("py-numpy@1.21.6:1.27", when="@1.11:", type=("build", "link", "run"))
+    depends_on("py-numpy@1.22.4:2", when="@1.12:", type=("build", "link", "run"))
+    depends_on("py-numpy@1.21.6:1.27", when="@1.11", type=("build", "link", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-scipy/package.py]: Same here

File: var/spack/repos/builtin/packages/py-scipy/package.py [hunk 3]
```diff
@@ -119,6 +120,9 @@ class PyScipy(PythonPackage):
         "when building with MSVC",
     )
 
+    # https://github.com/scipy/scipy/issues/19831
+    conflicts("openblas@0.3.26:", when="@:1.12")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-scipy/package.py]:
```suggestion
    conflicts("^openblas@0.3.26:", when="@:1.12")
```

Comment (liam-o-marsh) [reply var/spack/repos/builtin/packages/py-scipy/package.py]: done

File: var/spack/repos/builtin/packages/py-scipy/package.py (hunk 2 above)
Comment (liam-o-marsh) [reply var/spack/repos/builtin/packages/py-scipy/package.py]: done

File: var/spack/repos/builtin/packages/py-scipy/package.py (hunk 1 above)
Comment (liam-o-marsh) [reply var/spack/repos/builtin/packages/py-scipy/package.py]: done

**Review by adamjstewart — APPROVED** [review]
This looks fine to me. We could also add the patch if we want but I'm personally fine with waiting. Will give @rgommers time to review as well.

**Comment by alalazo** [comment]
Given the small change set, and the fact that a week has passed, I'll merge this.
